sparkly/testing.py

```diff
@@ -120,28 +147,132 @@ def tearDown(self):
             fixture.teardown_data()
 
-    def assertDataFrameEqual(self, actual_df, expected_data, fields=None, ordered=False):
+    def assertDataFrameEqual(self,
+                             actual_df,
+                             expected_data,
+                             fields=None,
+                             ordered=False,
+                             places=None):
         """Ensure that DataFrame has the right data inside.
 
         Args:
             actual_df (pyspark.sql.DataFrame|list[pyspark.sql.Row]): Dataframe to test data in.
             expected_data (list[dict]): Expected dataframe rows defined as dicts.
             fields (list[str]): Compare only certain fields.
             ordered (bool): Does order of rows matter?
+            places (int|None): Compare floats up to given decimal places.
         """
         if fields:
             actual_df = actual_df.select(*fields)
 
         actual_rows = actual_df.collect() if hasattr(actual_df, 'collect') else actual_df
         actual_data = [row.asDict(recursive=True) for row in actual_rows]
 
-        if ordered:
-            self.assertEqual(actual_data, expected_data)
-        else:
-            try:
-                self.assertCountEqual(actual_data, expected_data)
-            except AttributeError:
-                self.assertItemsEqual(actual_data, expected_data)
+        return self.assertEqual(
+            actual_data,
+            expected_data,
+            ignore_order=not ordered,
+            max_recursion_depth=1,
+            places=places,
+        )
+
+    def assertEqual(self,
+                    first,
+                    second,
+                    msg=None,
+                    ignore_order=False,
+                    max_recursion_depth=None,
+                    places=None,
+                    significant=False,
+                    equal_nan=True):
+        """Assert equal on steroids.
+
+        Extend this classic function signature to work better with
+        comparisons involving dataframes, rows, dictionaries and lists by
+        ignoring the order of mutable nested structures recursively and
+        rounding up floats when necessary.
+
+        Args:
+            first, second, msg: see unittest.TestCase.assertCountEqual
+            ignore_order (bool|False): ignore the order in all structures
+                that are inherently oredered (e.g., lists, mock calls).
+            max_recursion_depth (int|None): if ignore_order is true, do
+                ignore order up to this nesting level (exclusive). Setting
+                this to 0 is same to setting ignore_order=False, 1 means
+                ignore order only at the top level and so on. Default is
+                ignore order arbitrarily deep.
+            places (int|None): Decimal or significant digits to consider
+                during comparisons of floats.
+            significant (bool|False): If set, interpret places as # of
```

Comment:
changed it quite a bit, and added more explanations. hopefully it's cleaner now!